File: system-test/java/src/test/java/com/optivem/shop/systemtest/legacy/mod02/base/BaseRawTest.java
package com.optivem.shop.systemtest.legacy.mod02.base;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.microsoft.playwright.*;
import com.optivem.shop.systemtest.configuration.BaseConfigurableTest;
import com.optivem.shop.systemtest.configuration.Configuration;
import com.optivem.shop.testkit.common.Closer;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;

import java.net.http.HttpClient;
import java.util.UUID;

public class BaseRawTest extends BaseConfigurableTest {
    protected Configuration configuration;

    protected Playwright shopUiPlaywright;
    protected Browser shopUiBrowser;
    protected BrowserContext shopUiBrowserContext;
    protected Page shopUiPage;

    protected HttpClient shopApiHttpClient;
    protected HttpClient erpHttpClient;
    protected HttpClient taxHttpClient;

    protected ObjectMapper httpObjectMapper;

    @BeforeEach
    protected void setUpConfiguration() {
        configuration = loadConfiguration();
    }

    protected void setUpShopBrowser() {
        shopUiPlaywright = Playwright.create();

        var launchOptions = new BrowserType.LaunchOptions()
                .setHeadless(true);

        shopUiBrowser = shopUiPlaywright.chromium().launch(launchOptions);

        var contextOptions = new Browser.NewContextOptions()
                .setViewportSize(1920, 1080)
                .setStorageStatePath(null);

        shopUiBrowserContext = shopUiBrowser.newContext(contextOptions);
        shopUiPage = shopUiBrowserContext.newPage();
    }

    protected void setUpShopHttpClient() {
        shopApiHttpClient = HttpClient.newBuilder().version(HttpClient.Version.HTTP_1_1).build();
        if (httpObjectMapper == null) {
            httpObjectMapper = createObjectMapper();
        }
    }

    protected void setUpExternalHttpClients() {
        erpHttpClient = HttpClient.newBuilder().version(HttpClient.Version.HTTP_1_1).build();
        taxHttpClient = HttpClient.newBuilder().version(HttpClient.Version.HTTP_1_1).build();
        httpObjectMapper = createObjectMapper();
    }

    protected String getShopApiBaseUrl() {
        return configuration.getShopApiBaseUrl();
    }

    protected String getShopUiBaseUrl() {
        return configuration.getShopUiBaseUrl();
    }

    protected String getErpBaseUrl() {
        return configuration.getErpBaseUrl();
    }

    protected String getTaxBaseUrl() {
        return configuration.getTaxBaseUrl();
    }

    private ObjectMapper createObjectMapper() {
        var mapper = new ObjectMapper();
        mapper.registerModule(new JavaTimeModule());
        return mapper;
    }

    protected String createUniqueSku(String baseSku) {
        var suffix = UUID.randomUUID().toString().substring(0, 8);
        return baseSku + "-" + suffix;
    }

    @AfterEach
    void tearDown() {
        Closer.close(shopUiPage);
        Closer.close(shopUiBrowserContext);
        Closer.close(shopUiBrowser);
        Closer.close(shopUiPlaywright);
        Closer.close(erpHttpClient);
        Closer.close(taxHttpClient);
        Closer.close(shopApiHttpClient);
    }
}

File: system-test/typescript/tests/legacy/mod02/base/BaseRawTest.ts
import { randomUUID } from 'node:crypto';
import { chromium, type Browser, type BrowserContext, type Page } from 'playwright';
import { loadConfiguration, type TestConfig } from '../../../../config/configuration-loader.js';

export interface MyShopBrowser {
  browser: Browser;
  context: BrowserContext;
  page: Page;
}

export function getConfiguration(): TestConfig {
  return loadConfiguration();
}

export function getMyShopApiBaseUrl(config: TestConfig = getConfiguration()): string {
  return config.myShop.backendApiUrl;
}

export function getMyShopUiBaseUrl(config: TestConfig = getConfiguration()): string {
  return config.myShop.frontendUrl;
}

export function getErpBaseUrl(config: TestConfig = getConfiguration()): string {
  return config.externalSystems.erp.url;
}

export function getTaxBaseUrl(config: TestConfig = getConfiguration()): string {
  return config.externalSystems.tax.url;
}

export function createUniqueSku(baseSku: string): string {
  const suffix = randomUUID().substring(0, 8);
  return `${baseSku}-${suffix}`;
}

export async function setUpMyShopBrowser(): Promise<MyShopBrowser> {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({ viewport: { width: 1920, height: 1080 } });
  const page = await context.newPage();
  return { browser, context, page };
}

export async function tearDownMyShopBrowser(myShopBrowser: MyShopBrowser | null): Promise<void> {
  if (!myShopBrowser) return;
  await myShopBrowser.page.close().catch(() => {});
  await myShopBrowser.context.close().catch(() => {});
  await myShopBrowser.browser.close().catch(() => {});
}
